### PR TITLE
10.0 dev reload with inotify

### PR DESCRIPTION
**Pull request by Icallhimtest**
Add the alternative of inotify instead of watchdog to watch the addons path the server was started with.

Reason: watchdog spawns 2 threads per path to watch. When there are a lot of addons paths, this can become too costly. With inotify we watch all the repositories in a single thread.

https://github.com/dsoprea/PyInotify

installation:
        pip install inotify

This PR is a bit more dirty than what's suggested in https://github.com/odoo-dev/odoo/tree/10.0-dev-reload-dve but a lot more efficient.

### CODE REVIEW

File: odoo/service/server.py [hunk 1]
```diff
@@ -941,21 +1000,28 @@ def start(preload=None, stop=False):
...
+    if 'reload' in config['dev_mode'] and not odoo.evented:
+        if inotify and not odoo.evented:
+            watcher = FSWatcherInotify()
+            watcher.start()
+        elif watchdog:
             watcher = FSWatcher()
             watcher.start()
         else:
-            _logger.warning("'watchdog' module not installed. Code autoreload feature is disabled")
+            if os.name == 'posix':
```
Comment (KangOl) [inline odoo/service/server.py]: there is not `inotify` on osx. watchdog is still needed

File: odoo/service/server.py [hunk 2]
```diff
@@ -941,21 +1000,28 @@ def start(preload=None, stop=False):
         server = ThreadedServer(odoo.service.wsgi_server.application)
 
     watcher = None
-    if 'reload' in config['dev_mode']:
-        if watchdog:
+    if 'reload' in config['dev_mode'] and not odoo.evented:
+        if inotify and not odoo.evented:
```
[the remark this answers is not in the document]
Comment (Icallhimtest) [reply odoo/service/server.py]: whoops, indeed

File: odoo/service/server.py [hunk 3]
```diff
@@ -145,12 +156,60 @@ def dispatch(self, event):
...
+    def __init__(self):
+        self.started = False
+        self.watcher = None
+        # ignore warnings from inotify in case we have duplicate addons paths.
+        inotify.adapters._LOGGER.setLevel(logging.ERROR)
+        for path in odoo.modules.module.ad_paths:
+            if not self.watcher:
+                self.watcher = InotifyTree(path, mask=INOTIFY_LISTEN_EVENTS, block_duration_s=1.0)
+            else:
+                self.watcher._InotifyTree__load_tree(path)
```
[the remark this answers is not in the document]
Comment (Icallhimtest) [reply odoo/service/server.py]: Yeah, But this avoids having to iterate through different InotifyTree and have all the events on a single fd, which, on my 500 addons-paths test prevents a constant 2% CPU usage to iterate on every path every second.

File: odoo/service/server.py [hunk 4]
```diff
@@ -145,12 +156,60 @@ def dispatch(self, event):
...
+            _logger.info('Watching addons folder %s', path)
+
+    def run(self):
+        _logger.info('AutoReload watcher running with inotify')
+        while self.started:
+            for event in self.watcher.event_gen(timeout_s=0, yield_nones=False):
+                (_, type_names, path, filename) = event
+                if filename.endswith('.py') and not filename.startswith('.~') and 'IN_ISDIR' not in type_names:
```
[the remark this answers is not in the document]
Comment (Icallhimtest) [reply odoo/service/server.py]: will do

**Comment by Icallhimtest** [comment]
@KangOl  @antonylesuisse 
The `FSWatcherInotify` runs fine for the expected scenarios, but a few questions remain:

1) Inotify behaves a bit differently when moving folders containing files than watchdog. Watchdog will give the notifications for each file in the folder, while inotify only gives the folder move event (I believe because the files get created before inotify gets a chance to add the folder to its watchlist, not sure how watchdog handles this). The question becomes:

a) Do we automatically reload upon directory moves?
b) Do we os.walk the directory in search of python files and reload only if it contains one & it compiles?
c) Do we ignore them?  <-- current implementation

2) Current watchdog implementation ignores when we delete or move a file away from the folder, even a python file.

Is this intended behaviour? Current inotify implementation copies this. It doesn't listen to `IN_DELETE, IN_DELETE_SELF, IN_MOVE_SELF, IN_MOVED_FROM` events.  ( http://man7.org/linux/man-pages/man7/inotify.7.html )

3) Is it ok to use the `__load_tree` method from the `InotifyTree` object? Its signature hasn't changed since its addition 4 years ago. https://github.com/dsoprea/PyInotify/blame/master/inotify/adapters.py#L342

**Comment by KangOl** [comment]
On 15 Mar 2019, at 12:47, Denis Vermylen wrote:


@KangOl  @antonylesuisse
The `FSWatcherInotify` runs fine for the expected scenarios, but a few questions remain:

1) Inotify behaves a bit differently when moving folders containing files than watchdog. Watchdog will give the notifications for each file in the folder, while inotify only gives the folder move event (I believe because the files get created before inotify gets a chance to add the folder to its watchlist, not sure how watchdog handles this). The question becomes:

a) Do we automatically reload upon directory moves?
b) Do we os.walk the directory in search of python files and reload only if it contains one & it compiles?
c) Do we ignore them?  <-- current implementation




It's better to have the same behaviour with both packages. I would go with b) then.


2) Current watchdog implementation ignores when we delete or move a file away from the folder, even a python file.

Is this intended behaviour? Current inotify implementation copies this. It doesn't listen to `IN_DELETE, IN_DELETE_SELF, IN_MOVE_SELF, IN_MOVED_FROM` events.  ( http://man7.org/linux/man-pages/man7/inotify.7.html )




You choose. You can keep current status or implement it.


3) Is it ok to use the `__load_tree` method from the `InotifyTree` object? Its signature hasn't changed since its addition 4 years ago. https://github.com/dsoprea/PyInotify/blame/master/inotify/adapters.py#L342





There is actually InotifyTrees (with a s) that take a list of paths.

**Comment by Icallhimtest** [comment]
@KangOl whoops, good thing you saw the InotifyTree**s** class, I completely missed it. It'll be nice & clean this way.

For point 1, looking at watchdog, they generate all the events manually when a directory is created:

https://github.com/gorakhargosh/watchdog/blob/master/src/watchdog/observers/inotify.py#L156
https://github.com/gorakhargosh/watchdog/blob/master/src/watchdog/events.py#L612

I'm hesitating between submitting a PR to inotify adding this feature somewhere here https://github.com/dsoprea/PyInotify/blame/master/inotify/adapters.py#L287
or simply doing it in Odoo.

**Comment by KangOl** [comment]
From the reactivity of maintainer of `pyinotify`, it's better to handle it inside odoo, even if a PR is made upstream.

**Comment by Icallhimtest** [comment]
@KangOl alrighty,

1b) applied & tested on my end
2) kept same as watchdog
3) done

Should be good to go.

**Comment by antonylesuisse** [comment]
Make sure it also works for editor that write in place.

**Comment by Icallhimtest** [comment]
@antonylesuisse Yeah, it does.

It listens to IN_WRITE or IN_MOVED_TO events. (+ IN_CREATE)

Both swapping or writing in place are covered by this.

**Comment by Icallhimtest** [comment]
@KangOl Want any more changes or are we good to go here?

**Pull request closed without merging**